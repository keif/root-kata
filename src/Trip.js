const HIGHWAY_SPEED = 50;

class Trip {
    constructor(startTime, endTime, milesDriven) {
        this._startTime = startTime;
        this._endTime = endTime;
        this._totalMilesDriven = Number(milesDriven);
        this._duration = this._calcTotalTime();
    }

    _calcTotalTime() {
        const [startHour, startMin] = this._convertTimeToNumber(this._startTime);
        const [endHour, endMin] = this._convertTimeToNumber(this._endTime);

        return endHour + endMin / 60 - (startHour + startMin / 60);
    }

    _convertTimeToNumber(drivingTimes) {
        return drivingTimes.split(":").map(num => Number(num));
    }

    getDuration() {
        return this._duration;
    }

    getAverageSpeed() {
        return Math.round(this._totalMilesDriven / this._duration);
    }

    getMilesDriven() {
        return this._totalMilesDriven;
    }

    getIsHighway() {
        return this.getAverageSpeed() >= HIGHWAY_SPEED;
    }
}

export default Trip;
