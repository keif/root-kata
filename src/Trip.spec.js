import Trip from "./Trip";

let trip;
let tripDetails = {
    default: "07:15 07:45 17.3".split(" "),
    ave100mph: "07:00 08:00 100".split(" "),
    ave50mph: "07:00 08:00 50".split(" ")
};

describe("Trip Class", () => {
    describe("with a given valid trip", function() {
        beforeEach(() => {
            trip = new Trip(...tripDetails.default);
        });

        test("generates a trip object", () => {
            expect(trip).toMatchSnapshot();
        });

        test("getAverageSpeed returns the calculated average speed for the contructed trip", () => {
            expect(trip.getAverageSpeed()).toEqual(35);
        });

        test("getDuration returns the calculated drive time for the constructed trip", () => {
            expect(trip.getDuration()).toEqual(0.5);
        });

        describe("getIsHighway", () => {
            test("returns if the trip average speed exceeds or equals HIGHWAY_SPEED", () => {
                let tripIsHighway100mph = new Trip(...tripDetails.ave100mph);
                let tripIsHighway50mph = new Trip(...tripDetails.ave50mph);

                expect(trip.getIsHighway()).toEqual(false);
                expect(tripIsHighway100mph.getIsHighway()).toEqual(true);
                expect(tripIsHighway50mph.getIsHighway()).toEqual(true);
            });
        });

        test("getMilesDriven returns the calculated distance for the contructed trip", () => {
            expect(trip.getMilesDriven()).toEqual(17.3);
        });
    });
});
