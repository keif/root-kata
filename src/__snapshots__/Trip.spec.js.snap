// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trip Class generates a trip object 1`] = `
Trip {
  "_duration": 0.5,
  "_endTime": "07:45",
  "_startTime": "07:15",
  "_totalMilesDriven": 17.3,
}
`;
